dlock/dlock.go

```diff
@@ -0,0 +1,150 @@
+// Package dlock is a distributed lock to enable advanced synchronization for Mattermost Plugins.
+//
+// if you're new to distributed locks and Mattermost Plugins please read sample use case scenarios
+// at: https://community.mattermost.com/core/pl/bb376sjsdbym8kj7nz7zrcos7r
+package dlock
+
+import (
+	"context"
+
+	"sync"
+	"time"
+
+	pluginapi "github.com/lieut-data/mattermost-plugin-api"
+	"github.com/pkg/errors"
+)
+
+const (
+	// storePrefix used to prefix lock related keys in KV store.
+	storePrefix = "dlock:"
+)
+
+const (
+	// lockTTL is lock's expiry time.
+	lockTTL = time.Second * 15
+
+	// lockRefreshInterval used to determine how long to wait before refreshing
+	// a lock's expiry time.
+	lockRefreshInterval = time.Second
+
+	// lockTryInterval used to wait before trying to obtain the lock again.
+	lockTryInterval = time.Second
+)
+
+// API is a portion of pluginapi to keep locks' state and log errors.
+type API interface {
+	Set(key string, value interface{}, options ...pluginapi.KVSetOption) (bool, error)
+	Error(message string, keyValuePairs ...interface{})
+}
+
+// DLock is a distributed lock.
+type DLock struct {
+	// api used to keep locks' state and log errors.
+	api API
+
+	// key to lock for.
+	key string
+
+	// refreshCancel stops refreshing lock's TTL.
+	refreshCancel context.CancelFunc
+
+	// refreshWait is a waiter to make sure refreshing is finished.
+	refreshWait *sync.WaitGroup
+}
+
+// New creates a new distributed lock for key on given store with options.
+// think,
+//   `dl := New("my-key", store)`
+// as an equivalent of,
+//   `var m sync.Mutex`
+// and use it in the same way.
+func New(key string, api API) *DLock {
+	return &DLock{key: buildKey(key), api: api}
+}
+
+// Lock obtains a new lock.
+// ctx provides a context to locking. when ctx is cancelled, Lock() will stop
+// blocking and retries and return with error.
+// use Lock() exactly like sync.Mutex.Lock(), avoid missuses like deadlocks.
+func (d *DLock) Lock(ctx context.Context) error {
+	for {
+		isLockObtained, err := d.lock()
+		if err != nil {
+			return err
+		}
+		if isLockObtained {
+			return nil
+		}
+		afterC := time.After(lockTryInterval)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-afterC:
+		}
+	}
+}
+
+// TryLock tries to obtain the lock immediately.
+// err is only filled on system failure.
+func (d *DLock) TryLock() (isLockObtained bool, err error) {
```

Comment:
1/5 In general I don't see the necessity to name the return values. The go doc comment should make clear, what the method does.